### ISSUE TITLE
Missing LICENSE and COPYRIGHT statements

### ISSUE THREAD

**Issue by jwnimmer-tri** [RobotLocomotion/lcmtypes#2]
Code is unusable without LICENSE and COPYRIGHT.  Please figure out the existing licensing of all of the copied code, and then update the files here to match (as derived works).

**Comment by mwoehlke-kitware** [comment]
The license for the libbot types appears to be LGPL. I guess the OH folks that forked that didn't pay much attention to such matters, so I have no idea what license applies to their contributions?

IANAL but my guess is we can't fix this without confirmation from the stakeholders... I'll try to ping them...

**Comment by jwnimmer-tri** [comment]
That's awful.  Good luck!

**Comment by patmarion** [comment]
For what it's worth, the openhumanoids code and all components is licensed with the 3 clause bsd license:

https://github.com/openhumanoids/oh-distro/blob/master/LICENSE.txt

**Comment by mwoehlke-kitware** [comment]
> For what it's worth, the openhumanoids code and all components is licensed with the 3 clause bsd license

I don't think we can apply that without at least @ashuang's approval, since the original libbot types appear to be under LGPLv3. (In fact, if you (i.e. OH) are applying that to OH's LCM types, including those from libbot, you are probably violating @ashuang's copyright, unless permission was obtained.)

p.s. @patmarion, you are one of the stakeholders... did you get my mail? Do you agree to license your contributions under LGPLv3 and/or BSD 3-clause? (Also, if you have contact with any of the other stakeholders and can poke them to reply, that would be appreciated :smile:.)

**Comment by mwoehlke-kitware** [comment]
Okay, trying again to identify the stakeholders...
- [x] @ashuang
  - [x] LGPLv3+
  - [x] BSD 3-clause
- [x] @iamwolf (a.k.a. @wxmerkt)
  - [x] LGPLv3+
  - [x] BSD 3-clause
- [x] @rdeits
  - [x] LGPLv3+
  - [x] BSD 3-clause
- [x] @mauricefallon
  - [x] LGPLv3+
  - [x] BSD 3-clause
- [x] @mrwalter
  - [x] LGPLv3+
  - [x] BSD 3-clause

If you are named in this comment, please indicate if you approve licensing the code in this repository under [LGPL, version 3.0 or later](https://opensource.org/licenses/LGPL-3.0) and/or [BSD 3-clause](https://opensource.org/licenses/BSD-3-Clause). Please also indicate the appropriate copyright holder (e.g. you as an individual, or the organization under which your contributions were made).

For the record, I currently have these copyrights:
- 2012 Albert Huang (original libbot2 types)
- 2012 Matthew Walter (original libbot2 types, presumably)
- 2015 MIT DRC Team @ CSAIL (stuff OH added)
- 2016 Maurice Fallon
- 2016 Wolfgang Merkt
- 2016 Robot Locomotion Group @ CSAIL (more recent changes from Drake, including those made by Kitware)
- 2016 University of Edinburgh

~~(@mauricefallon, can you confirm if your changes from 2016 do indeed bump the MIT DRC date from 2015 to 2016, as shown above?)~~

**Comment by mauricefallon** [comment]
I'm happy with either license.

I'm not sure what your question regarding 2015 and 2016 means. I have not been an MIT employee since Dec 2014. You seem to be being ultra careful, in which case you should add University of Edinburgh - which is where @iamwolf and I work and have been since Jan 2015.

Also in theory all open source requests at MIT (including Albert's work) should go through the MIT Technology Licensing Office. Can-of-worms.

**Comment by ashuang** [comment]
I approve of both licenses. Some of my work is copyright myself, and some
is probably copyright MIT.

Albert

**Comment by rdeits** [comment]
I approve of both licenses. All of my work on this project has either been through the MIT DRC team or through the Locomotion group.

**Comment by mwoehlke-kitware** [comment]
> I'm not sure what your question regarding 2015 and 2016 means. I have not been an MIT employee since Dec 2014. You seem to be being ultra careful, in which case you should add University of Edinburgh - which is where @iamwolf and I work and have been since Jan 2015.

Okay, updated. I "reverted" the year change to MIT DRC Team. Do you know if they have contributed any changes since 2015?

> Also in theory all open source requests at MIT (including Albert's work) should go through the MIT Technology Licensing Office.

Are any of you aware of contributions to either libbot2 or originating from MIT but _not_ from OH to which this would apply? (The libbot2 stuff is previously released under LGPLv3+; the OH stuff is already released under BSD.)

> Some of my work is copyright myself, and some is probably copyright MIT.

If any of the MIT work needs authorship other than I listed above, please let me know. Otherwise, NP :smile:.

**Comment by mauricefallon** [comment]
no changes since 2015. MIT DRC team has been dissolved

and no other contributions

**Comment by mwoehlke-kitware** [comment]
Okay, thanks!

Waiting on @iamwolf. (Those contributions, all 2016, are presumably attributed to "University of Edinburgh"?)

**Comment by wxmerkt** [comment]
Hey @mwoehlke-kitware,
I am an individual/myself (student; IP here is held by students unless assigned, which isn't the case). Happy with either LGPLv3+ or 3-clause BSD.

Also, I've been trying to push forward your modernisation with added backwards compatibility in https://github.com/openhumanoids/bot_core_lcmtypes/pull/20 - thanks so much for your contribution. I'd really love to see if there's a way to not diverge on the communication/message level, I'd really appreciate that (and have a strong interest in it) - and also hope this isn't closing the door after horse has already bolted. It's more than unfortunate and unsatisfying that response and action has taken so long, I apologise as I couldn't be the sole student to pick up maintenance, assessment, and integration. That being said, if we can find a way to continue to work with shared types and not diverge, I am happy to put dev cycles towards it and also maintain two branches for the two build systems and apply changes across them if required.

Best
Wolfgang

**Comment by mwoehlke-kitware** [comment]
> I am an individual/myself (student; IP here is held by students unless assigned, which isn't the case). Happy with either LGPLv3+ or 3-clause BSD.

Thanks. I updated the copyrights list. Please let me know if didn't do it right :smile:.

@mauricefallon, should your contributions be assigned to UoE, yourself, or both?

**Comment by mwoehlke-kitware** [comment]
> That being said, if we can find a way to continue to work with shared types and not diverge, I am happy to put dev cycles towards it and also maintain two branches for the two build systems and apply changes across them if required.

This needs to be discussed somewhere else, though I'm not certain what would be the correct forum. There are a couple issues:
- Drake needs fast turn-around. We can't wait on other projects to accept integrations, both for build system changes and, especially, for the addition of new types. While we may in some cases be able to maintain a fork and keep trying to push changes upstream, openhumanoids/bot_core_lcmtypes#20 is indeed a case of "closing the door after horse has already bolted". (Note especially that OH initially _rejected_ the changes outright.)
- I'd really like to stop misrepresenting new types as belonging to libbot. This includes doing so retroactively (i.e. biting the bullet and moving types that were never part of libbot to a new namespace). As I'm now in a position to help maintain libbot, or at least select components thereof, it may be possible to split the libbot types, and any DRC types that we want to make _officially_ part of libbot, into a separate repo. This would still rely on modern LCM (i.e. master, at least until 1.4.0 is released), but would be better positioned to take over from the monolithic libbot repo without other disruption. (Especially as it seems the libbot types have been pretty stable for some time.) However, I'm not sure if Drake will accept having multiple repositories for LCM types, even if all of them are effectively under our control.

TBH, the hope/desire however was/is that this repo would become the new primary repo, and other users would slowly migrate.

Rather than discuss here, I would request that someone knowledgeable propose a more appropriate forum, bearing in mind that both Drake and OH are major stakeholders.

**Comment by RussTedrake** [comment]
@wxmerkt -- let me make sure I understand.  I believe that the message types will not change (we might add to them).  the build system is changing.  My hope is that this happens quickly and that you could switch over to using this as a cmake external so we stay in sync with each other.  Is that overly simplistic/optimistic?

@mwoehlke-kitware , @RobotLocomotion/tri -- note that actually changing the types has significant additional cost to existing users because it means we have to use an old revision in order to inspect old logs.  Best practices are to change LCM types only when absolutely necessary.

**Comment by mwoehlke-kitware** [comment]
The message types _already changed_ (specifically, the namespace changed; see #1). Per previous comments, I think this is TRTTD, at least for some types, because we've historically abused the bot_core namespace for non-libbot types.

Can we _please_ take this off topic conversation to a more appropriate venu? Is there a mailing list that makes sense for this and is accessible to all stakeholders, or do I need to open a new issue? (I'd prefer to not have what seems likely to turn into a lengthy discussion on github.)

**Comment by mauricefallon** [comment]
> @mauricefallon, should your contributions be assigned to UoE, yourself, or both?

Both.

**Comment by rdeits** [comment]
As far as I know, changing the namespace of the types does not actually change their LCM fingerprint, and does not break backwards compatibility. For example, we have at times had `pronto::` LCM types with exactly the same fields as `drc::` LCM types, and have been able to treat them as identical.

**Comment by mwoehlke-kitware** [comment]
> As far as I know, changing the namespace of the types does not actually change their LCM fingerprint, and does not break backwards compatibility.

Ah, cool. (Well, it breaks SC and BC, but oh well; if they're wire compatible that likely covers the really critical use cases that folks are worrying over. I should note also I don't think there are any plans at this time to change the actual data formats, at least not for the more established types.)

**Comment by ashuang** [comment]
LCM fingerprint is invariant to changes in package name and type name.
e.g., the following have identical fingerprints and wire format:

package abc;
struct vec2d {
  double x;
  double y;
}

package def;
struct point2d {
  double x;
  double y;
}

**Comment by mwoehlke-kitware** [comment]
> > @mauricefallon, should your contributions be assigned to UoE, yourself, or both?
> 
> Both.

Thanks. I've updated the list accordingly, and I think that covers everyone. I will start preparing a PR to add a `COPYRIGHT` and `LICENSE`. If anyone is aware of anything I have missed, please speak up ASAP!

Thanks again, everyone!

**Comment by david-german-tri** [comment]
I've created #3 to track the ecosystem strategy and protocol compatibility discussion.  I agree with @mwoehlke-kitware it will be clearer to separate that from this licensing discussion.

**Comment by jwnimmer-tri** [comment]
Seriously, guys?  This library got added to Drake but doesn't have a license file yet?  Can we _please_ get licensing sorted out ASAP?

**Comment by jwnimmer-tri** [comment]
@mwoehlke-kitware @sammy-tri reping.  What's the deal with getting this code marked with copyright ownership, and an open-source license grant?

**Comment by sammy-tri** [comment]
The initial code was copied from openhumanoids/oh-distro, which has this license: https://github.com/openhumanoids/oh-distro/blob/master/LICENSE.txt

Do you think that would be suitable to copy into this repository?

**Comment by jwnimmer-tri** [comment]
Sure, files that are bitwise identical to code under that license can be used and redistributed under those terms (BSD-3-Clause).  We can mark them as such in our fork.  If there were subsequent modifications, we'll need to understand the license grant of those modifications.

**Comment by sammy-tri** [comment]
There have been subsequent modifications.  If you're implying that there's additional understanding required to take code licensed under BSD-3-Clause, modify it, and continue licensing it under BSD-3-Clause, you want to talk to a lawyer, not me.  If you're implying I don't have the authority to claim the files which I modified are still valid to license under BSD-3-Clause, I'm not sure how to proceed from here.

**Comment by jwnimmer-tri** [comment]
> If you're implying that there's additional understanding required to take code licensed under BSD-3-Clause, modify it, and continue licensing it under BSD-3-Clause, you want to talk to a lawyer, not me.

Well, if the submitters of the changes agreed to contribute their modifications under BSD-3-Clause, then we are fine.  (If its only TRI folks, then we are fine, because TRI owns our changes, not us.)  If other contributors did not grant a license for their (non-trivial, i.e. copyrightable) changes, then nobody can (legally) do anything with those changes and we have to either approach them to grant a license, or force-push-remove their changes from the tree.

This is all a bit abstract.  Should I do a sit-down at some point and shoulder-surf the history in question, to help sort this out?

**Comment by sammy-tri** [comment]
The full history has only two committers (and one is me).  Let's do a quick sit-down when we're both in the same place, I suspect we can do this more quickly f2f.

**Comment by mwoehlke-kitware** [comment]
FYI: My last status (as of October :cry:) is that Matthew Walter was claiming copyright on "some of those LCM types", but never pinged this issue (which I requested for recording purposes) with license approval and copyright statement. AFAIUI he did approve of the licensing, however, but I'd prefer to have an "official" record.

AFAIK, all other stakeholders are identified in my list, above, and have approved using either BSD-3 or LGPL.

> The full history has only two committers [...]

The full history must be traced back through libbot, as some of the types came from there. I'm pretty sure there are more than two contributors. IIRC some of said history is not even available any more.

Note that the issue is that it isn't clear that OH had permission to distribute under BSD-3 those types which originated in libbot. This issue has recorded permission from all stakeholders *except Matthew Walter* to do so.

**Comment by mwoehlke-kitware** [comment]
Addendum: the previous comment applies to the types in bot_core_types (which is also missing proper attribution). Types unique to the RobotLocomotion repo shouldn't be in the same mess, but I'd like to get the OH repo fixed, also...

**Comment by sammy-tri** [comment]
From f2f:  Create PR adding license file, Pat to ack as the other committer to the repo.  Add documentation (see somewhere in Drake) that PR contributors license their contributions with the same license as the project.  (I typed that before seeing the most recent updates about bot_core_lcmtypes)

**Comment by mrwalter** [comment]
Sorry to be piping in so late. I am fine with both licenses. Some copyrights are to me while others are to MIT.

**Comment by mwoehlke-kitware** [comment]
Thanks, @mrwalter! Do you know what year your contributions were made?

**Comment by mrwalter** [comment]
I'd say starting in 2012

**Comment by mwoehlke-kitware** [comment]
Okay, though the *end* date is most useful. This is for the copyright statement; for now, I'll put 2012, but let me know if that should be changed. Thanks again!

**Comment by mwoehlke-kitware** [comment]
Okay! I believe https://github.com/RobotLocomotion/lcmtypes/issues/2#issuecomment-251760790 now reflects the complete list of needed approvals and attributions for the combined types in libbot2, OH/bot_core_types, and RL/lcmtypes. Not all attributions may apply to all repos, but at any rate, I do believe all our approvals are now in order.

Please ping this issue if any of the attributions need adjusting.

Thanks again, all!

**Comment by sammy-tri** [comment]
Fixed in #9

**Comment by mwoehlke-kitware** [comment]
Note: At the time of this discussion, this repository was a fork of openhumanoids/bot_core_lcmtypes@6f8d14dc2f810aaf995c640518e6c3431179de98 (and was subsequently "rebooted" per #3). Accordingly, this discussion and the licensing approvals are pertinent to openhumanoids/bot_core_lcmtypes#33.
